core/mempool/src/pool_task.rs

```diff
@@ -0,0 +1,167 @@
+            tokio::spawn(
+                payload_request_tx
+                    .clone()
+                    .send(PayloadRequest::BlockProposal(authority_id, hash))
+                    .map(|_| ())
+                    .map_err(|e| warn!(target: "pool", "Error sending message: {}", e)),
+            );
+
+            pool1.add_pending(authority_id, hash);
+        } else {
+            let send_confirmation = control_tx1
+                .clone()
+                .send(Control::PayloadConfirmation(authority_id, hash))
+                .map(|_| ())
+                .map_err(|_| error!("Fail sending control signal to nightshade"));
+            tokio::spawn(send_confirmation);
+        }
+        future::ok(())
+    });
+    tokio::spawn(task);
+
+    // Make announcements of new payloads created from this node.
+    let pool2 = pool.clone();
+    let task = Interval::new_interval(payload_announce_period)
+        .for_each(move |_| {
+            if let Some((authority_id, payload)) = pool2.prepare_payload_announce() {
+                info!(
+                    "[{}] Payload confirmed from {}",
+                    pool2.authority_id.read().expect(""),
+                    authority_id
+                );
+                tokio::spawn(
+                    payload_announce_tx
+                        .clone()
+                        .send((authority_id, payload))
+                        .map(|_| ())
+                        .map_err(|e| warn!(target: "pool", "Error sending message: {}", e)),
+                );
+            }
+            future::ok(())
+        })
+        .map_err(|e| error!("Timer error: {}", e));
+    tokio::spawn(task);
+
+    // Receive payload and send confirmation signal of unblocked payloads.
+    let pool3 = pool.clone();
+    let control_tx2 = control_tx.clone();
+    let task = payload_response_rx.for_each(move |payload_response| {
+        if let Err(e) = match payload_response {
+            PayloadResponse::General(payload) => pool3.add_payload(payload),
+            PayloadResponse::BlockProposal(authority_id, payload) => {
+                pool3.add_payload_snapshot(authority_id, payload)
+            }
+        } {
+            warn!("Failed to add incoming payload: {}", e);
+        }
+
+        for (authority_id, hash) in pool3.ready_snapshots() {
+            let send_confirmation = control_tx2
+                .clone()
+                .send(Control::PayloadConfirmation(authority_id, hash))
+                .map(|_| ())
+                .map_err(|_| error!("Fail sending control signal to nightshade"));
+            tokio::spawn(send_confirmation);
+        }
+
+        future::ok(())
+    });
+    tokio::spawn(task);
+
+    let pool4 = pool.clone();
+    let task = mempool_control_rx.for_each(move |control| {
+        pool4.reset(control.clone());
+        let ns_control = match control {
+            MemPoolControl::Reset {
+                owner_uid,
```

Comment:
The following fields can be probably put in their own struct: `owner_uid,
	                block_index,
	                public_keys,
	                owner_secret_key,
	                bls_public_keys,
	                bls_owner_secret_key,`